### PR TITLE
feat: enforce AWS expiry on data lookups

### LINKED ISSUE #1051: Transition to new router table that uses TTL instead of rotation
Convert autopush to use a TTL based router table to reduce DB writes. 
requires #1049

### PR DESCRIPTION
**Pull request by jrconlin**
Closes #1051

### CODE REVIEW

File: autopush/db.py [hunk 1]
```diff
@@ -654,16 +659,20 @@ def fetch_timestamp_messages(
...
             Limit=limit
         )
+        now = int(time.time())
         notifs = [
             WebPushNotification.from_message_table(uaid, x) for x in
-            response.get("Items")
+            response.get("Items") if x.get('expiry', now) >= now
         ]
+        if response.get("Items") and not notifs:
+            print "here"
```
Comment (bbangert) [inline autopush/db.py]: Huh?

File: autopush/db.py [hunk 2]
```diff
@@ -523,6 +523,7 @@ def all_channels(self, uaid):
         # Note: This only returns the chids associated with the UAID.
         # Functions that call store_message() would be required to
         # update that list as well using register_channel()
+        # TODO: once expiry is properly integrated, use FilterExpression
```
Comment (bbangert) [inline autopush/db.py]: What does that mean?

File: autopush/router/gcm.py [hunk 3]
```diff
@@ -115,6 +115,8 @@ def _route(self, notification, uaid_data):
         try:
             gcm = self.gcm[router_data['creds']['senderID']]
             result = gcm.send(payload)
+        except RouterException:
+            raise  # pragma nocover
```
Comment (bbangert) [inline autopush/router/gcm.py]: Did this somehow get dropped?

File: autopush/db.py (hunk 2 above)
Comment (jrconlin) [reply autopush/db.py]: Ideally, we want to use FilterExpression in our calls to DynamoDB. Unfortunately, while really efficient, it doesn't handle our need for backwards compatibility. We have some records that don't have an expiry yet set, so those will not be fetched.

Once the expiry has been put in place for several months, and we can safely assume most current records have it, then we can remove the code I've put in place to handle records that do not have an expiry set.

File: autopush/db.py (hunk 1 above)
Comment (jrconlin) [reply autopush/db.py]: crap, debug. sorry.

File: autopush/router/gcm.py (hunk 3 above)
Comment (jrconlin) [reply autopush/router/gcm.py]: This may be a merge error.

Comment (bbangert) [reply autopush/router/gcm.py]: Maybe we should have a test that triggers this to avoid future merge errors?

File: autopush/db.py (hunk 2 above)
Comment (bbangert) [reply autopush/db.py]: Ah, perhaps filing an issue for that now, and linking it here would help. We could then mark the issue blocked so we can remember when we should remove it.

File: autopush/tests/test_integration.py [hunk 4]
```diff
@@ -1005,7 +1005,7 @@ def test_ttl_batch_expired_and_good_one(self):
         data2 = str(uuid.uuid4())
         client = yield self.quick_register()
         yield client.disconnect()
-        for x in range(0, 12):
```
Comment (pjenvey) [inline autopush/tests/test_integration.py]: was this change necessary here? IIRC the total messages churned through in these batch tests was > 10 to thoroughly test the fetch_messages call with its default limit=10

Comment (jrconlin) [reply autopush/tests/test_integration.py]: yes, the problem was that we would load up with the 10 expired messages and never get to the one that wasn't expired. That's highlighted now that we're actually expiring messages. The next test checks batching.

File: autopush/db.py (hunk 2 above)
Comment (jrconlin) [reply autopush/db.py]: 👍 https://github.com/mozilla-services/autopush/issues/1073

Comment (bbangert) [reply autopush/db.py]: Oh, I meant referring to the issue in the code itself.

**Comment by jrconlin** [comment]
Design assumption flawed. closing.
